### PR TITLE
New: add no-async-promise-executor rule (fixes #10217)

### LINKED ISSUE #10217: [Feature request] disallow async promise executor
**Please describe what the rule should do:**

Promise executors must not be (technically, they can if you know what you're doing)  `async` else bad side effects happen, such as throwing (seemingly) sync errors `(A)` no longer rejecting the promise (thus unable to be caught by `(B)`), because the async function is executed in different frame:

```js
new Promise( async r => {
    throw new Error("aaa"); // (A)
})
.catch(() => {}) // (B)
```

Why is this different from not catching any potential async errors in a non-`async` executor? It isn't -- but, the above example gives off a feeling like you're throwing a sync error, while it's not, and for people new to async/await this may not be evident.

More info in this [SO answer](https://stackoverflow.com/a/43083793/927631).

Thus, I propose a rule that forbids `async` [*executor* function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
new Promise( async function (resolve, reject) {
    // any code
});
```

```js
new Promise( async (resolve, reject) => {
    // any code
});
```

```js
const executor = async function (resolve, reject) {
    // any code
};
new Promise( executor );
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It seems like to become common-enough error when async/await becomes more mainstream to include it in core.

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] New rule (https://github.com/eslint/eslint/issues/10217)

**What changes did you make? (Give an overview)**

This implements the `no-async-promise-executor` rule, as described in https://github.com/eslint/eslint/issues/10217.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

### CODE REVIEW

File: docs/rules/no-async-promise-executor.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+The executor function can also be an `async function`. However, this is usually a mistake, for a few reasons:
+
+* If an async executor function throws an error, the error will be lost and won't cause the newly-constructed `Promise` to reject. This could make it difficult to debug and handle some errors.
+* If a Promise executor function is using `await`, this is usually a sign that it is not actually necessary to use the `new Promise` constructor, or the scope of the `new Promise` constructor can be reduced.
+
+## Rule Details
+
+This rule aims to...
```
Comment (platinumazure) [inline docs/rules/no-async-promise-executor.md]: Needs more 😄

File: lib/rules/no-async-promise-executor.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "disallow using an async function as a Promise executor",
+            category: "Fill me in",
```
Comment (platinumazure) [inline lib/rules/no-async-promise-executor.js]: This should probably be Possible Errors

File: lib/rules/no-async-promise-executor.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        },
+        fixable: null,
+        schema: []
+    },
+
+    create(context) {
+        return {
+            "NewExpression[callee.name='Promise'][arguments.0.async=true]"(node) {
+                context.report({
+                    node: context.getSourceCode().getFirstToken(node.arguments[0]),
```
Comment (platinumazure) [inline lib/rules/no-async-promise-executor.js]: I imagine this wouldn't point at the right token if the node is surrounded by parens?

Comment (not-an-aardvark) [reply lib/rules/no-async-promise-executor.js]: I think it still would be. I'll add a test.

File: docs/rules/no-async-promise-executor.md (hunk 1 above)
Comment (not-an-aardvark) [reply docs/rules/no-async-promise-executor.md]: Oops, I was aiming for concise documentation but I think I overdid it.

**Comment by not-an-aardvark** [comment]
> Wondering if the rule should allow non-global `Promise` references?

I think it's more common for people to do things like `const Promise = require('bluebird')` (i.e. create a `Promise` variable that has compatible semantics with the ES2015 `Promise`) than it is for people to create a `Promise` variable with different semantics. So I'm leaning towards having the rule treat non-global `Promise` references the same as it treats global ones.

File: lib/rules/no-async-promise-executor.js (hunk 3 above)
Comment (platinumazure) [reply lib/rules/no-async-promise-executor.js]: Well, I'm wondering if we should just use `context.getSourceCode().getFirstToken(node.arguments[0], { filter: t => t.value === "async" })` or similar to be absolutely safe.

**Pull request merged**